### PR TITLE
[WIP] time annotations for pytorch hook

### PR DESCRIPTION
**Pull request by NRauschmayr**
### Description of changes:
Adding time annotations for pytorch hook to measure time of forward/backward pass and time spent per operation on GPU. 

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Vikas-kum** [comment]
This needs to use TraceEvents writer from https://github.com/awslabs/sagemaker-debugger/pull/247/files

File: smdebug/pytorch/hook.py [hunk 1]
```diff
@@ -54,6 +57,14 @@ def __init__(
         self.worker = self._get_worker_name()
         set_hook(self)
 
+        self.forward_start = 0
+        self.backward_start = 0
+        self.buffer = {}
+        self.profiler_enabled = profiler_enabled
+        self.profiler_steps = profiler_steps
+        self.autograd_profiler_enabled = False
+        timeline_writer = self._maybe_get_timeline_writer()
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:66]: there's a _write_trace_events_summary() in core/hook.py. thought of wrapping that in a a more formal API like record_trace_event(). so all hooks could do a self.record_trace_event(). what do you think?

File: smdebug/pytorch/hook.py [hunk 2]
```diff
@@ -99,7 +110,7 @@ def _log_params(self, module):
             # self.logger.debug(
             # "Processing the global step {0} for parameter {1}".format(self.step, pname))
             self._save_for_tensor(tensor_name=pname, tensor_value=param.data)
-
+
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:113]: nit: extra line

File: smdebug/pytorch/hook.py [hunk 3]
```diff
@@ -118,6 +129,47 @@ def _prepare_collections(self):
...
+               op_name="forward",
+               #timestamp=time.time(),
+               duration=self.forward_end-self.forward_start,
+               step_num=self.step)
+            self.timeline_writer.write_trace_events(
+               op_name="backward",
+               #timestamp=time.time(),
+               duration=self.backward_end-self.backward_start,
+               step_num=self.step)
+            self.timeline_writer.flush()
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:143]: writer is flushed in close_Writers(). is it required here?

File: smdebug/pytorch/hook.py [hunk 4]
```diff
@@ -118,6 +129,47 @@ def _prepare_collections(self):
 
     # This hook is invoked by trainer prior to running the forward pass.
     def forward_pre_hook(self, module, inputs):
+        if self.profiler_enabled and self.forward_start != 0:
+            self.timeline_writer.write_trace_events(
+               op_name="forward",
+               #timestamp=time.time(),
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:135]: I saw the negative timestamp too. will check how to fix it in the timeline writer

File: smdebug/pytorch/hook.py [hunk 5]
```diff
@@ -118,6 +129,47 @@ def _prepare_collections(self):
...
+               op_name="backward",
+               #timestamp=time.time(),
+               duration=self.backward_end-self.backward_start,
+               step_num=self.step)
+            self.timeline_writer.flush()
+            if self.step in self.profiler_steps:
+                if not self.autograd_profiler_enabled:
+                   torch.autograd._enable_profiler(torch.autograd.ProfilerConfig(torch.autograd.ProfilerState.CUDA, False))
+                   self.autograd_profiler_enabled = True
+                   self.start_profiler = time.time() * 1000000
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:148]: you think we should save this 1000000 as a constant that can be used throughout smdebug?

File: smdebug/pytorch/hook.py [hunk 6]
```diff
@@ -131,15 +183,17 @@ def forward_pre_hook(self, module, inputs):
             self.prepared_collections = True
 
         self._increment_step()
-
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:134]: nit: add line back

File: smdebug/pytorch/hook.py [hunk 7]
```diff
@@ -54,6 +57,14 @@ def __init__(
         self.worker = self._get_worker_name()
         set_hook(self)
 
+        self.forward_start = 0
+        self.backward_start = 0
+        self.buffer = {}
```
Comment (vandanavk) [inline smdebug/pytorch/hook.py:62]: is buffer used elsewhere?

File: smdebug/pytorch/hook.py (hunk 4 above)
Comment (vandanavk) [reply smdebug/pytorch/hook.py:135]: fixed in the latest commit in timeline writer. you can now pass timestamp here

File: smdebug/pytorch/hook.py (hunk 3 above)
Comment (NRauschmayr) [reply smdebug/pytorch/hook.py]: I removed the function call

File: smdebug/pytorch/hook.py (hunk 4 above)
Comment (NRauschmayr) [reply smdebug/pytorch/hook.py]: Thanks for fixing it! I added the timestamp.

File: smdebug/pytorch/hook.py (hunk 5 above)
Comment (NRauschmayr) [reply smdebug/pytorch/hook.py:146]: Yes, that's a good idea. Shall we add it to smdebug/core/config_constants.py ?

File: smdebug/pytorch/hook.py (hunk 7 above)
Comment (NRauschmayr) [reply smdebug/pytorch/hook.py]: I forgot to remove it. Fixed it.

File: smdebug/pytorch/hook.py (hunk 5 above)
Comment (vandanavk) [reply smdebug/pytorch/hook.py:146]: yes, that would be good. I'll update my PR with same changes once you make it

File: smdebug/pytorch/hook.py (hunk 1 above)
Comment (vandanavk) [reply smdebug/pytorch/hook.py:65]: changed the API name @NRauschmayr . Please rebase your PR

File: smdebug/pytorch/hook.py (hunk 5 above)
Comment (vandanavk) [reply smdebug/pytorch/hook.py:146]: added this to config constant in PR 247.

**Pull request closed without merging**